[Frontend] Re-enable imported AST verification

hamishknight opened the pull request:
The reasons for disabling it appear to now be outdated. Re-enable it, and run it after the entire pipeline has finished to make sure we verify any decls deserialised during SIL/IRGen.

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test source compatibility

CodaFi commented on lib/FrontendTool/FrontendTool.cpp:
```diff
@@ -1190,6 +1190,18 @@ static bool emitAnyWholeModulePostTypeCheckSupplementaryOutputs(
+/// before or after LLVM depending on when the ASTContext gets freed.
+static void performEndOfPipelineActions(CompilerInstance &Instance) {
+  // Verify the AST for all the modules we've loaded.
+  Instance.getASTContext().verifyAllLoadedModules();
```
Can we assert `hasASTContext` here?

CodaFi commented on lib/Serialization/ModuleFile.cpp:
```diff
@@ -2959,9 +2959,22 @@ Identifier ModuleFile::getDiscriminatorForPrivateValue(const ValueDecl *D) {
+      if (gpd->getDeclContext() == getAssociatedModule())
+        continue;
+    }
+
```
Should this be folded into `shouldVerify`?

hamishknight replied:
@CodaFi We only want to skip the verification for deserialized generic params, and because the ASTVerifier is defined in libAST, we can't check for `SerializedASTFile` there. We could add an additional flag to the `ModuleDecl`, but given this whole thing is a hack in the first place, I figured it was better to just put it here. What do you think?

hamishknight replied:
Oh wait, I guess we could check the `FileUnitKind`...

hamishknight replied:
I ended up moving this check into the verification function for `GenericTypeParamDecl`

hamishknight commented:
@swift-ci please test source compatibility Release

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test source compatibility

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test source compatibility

hamishknight commented:
@swift-ci please test Windows

hamishknight commented:
@swift-ci please test Windows

hamishknight commented:
@swift-ci please test source compatibility Release

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test Linux

hamishknight commented:
@swift-ci please test Linux

hamishknight committed:
[Frontend] Factor out performEndOfPipelineActions

hamishknight committed:
[Frontend] Re-enable imported AST verification

The reasons for disabling it appear to now be
outdated. Re-enable it, and run it after the
entire pipeline has finished to make sure we
verify any decls deserialised during SIL/IRGen.

hamishknight committed:
[Frontend] Set "failed to load" bit if stdlib fails

Set the "failed to load" flag on the main module
if the stdlib fails to load, as we'll bail without
adding any source files to the module.

hamishknight committed:
[Frontend] Set "failed to load" for failed partial module

If a partial module fails to load, the main module
may be left without files, so let's set the failed
to load bit.

hamishknight committed:
Avoid verifying un-parented GenericTypeParamDecls

When deserializing, we delay parenting
GenericTypeParamDecls until we install the
GenericParamList on the parent declaration, using
a dummy parent of the ModuleDecl until this
happens.

However when deserializing GenericTypeParamTypes,
we don't necessarily get the opportunity to
deserialize the parent, leaving them with the
dummy parent.

We should probably consider not deserializing
GenericTypeParamDecls at all, we currently only do
it to preserve the sugar of the name, this could
be serialized on the type instead.

For now though, adjust the dummy parent to be the
SerializedASTFile, and tell the ASTVerifier to
skip verification for such generic params.

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test Windows

hamishknight commented:
@CodaFi I noticed we should also probably be setting the "failed to load" bit if we fail to load a partial module. How does this look?

CodaFi commented:
@swift-ci clean smoke test macOS platform

CodaFi commented:
@swift-ci clean smoke test Linux platform

CodaFi commented:
@swift-ci clean smoke test Linux platform

CodaFi commented on lib/AST/ASTVerifier.cpp:
```diff
@@ -2732,7 +2732,18 @@ class Verifier : public ASTWalker {
-      if (!GTPD->getDeclContext()->isInnermostContextGeneric()) {
+
+      // Skip verification of deserialized generic param decls that have the
+      // the file set as their parent. This happens when they have not yet had
```
Just noticed the double "the".

hamishknight replied:
Ak, good spot!
